File: src/levels/L00-tutorial/ts/repaired.ts
type Order = { drink: string }

class OrderTaker {
  constructor(private cashier: Cashier) {}

  // region: OrderTaker#take
  take(drink: string) {
    this.cashier.charge({ drink })
  }
  // endregion
}

class Cashier {
  constructor(private barista: Barista) {}

  // region: Cashier#charge
  charge(order: Order) {
    console.log(`Cobrado: ${order.drink}`)
    this.barista.prepare(order)
  }
  // endregion
}

class Barista {
  constructor(private counter: Counter) {}

  // region: Barista#prepare
  prepare(order: Order) {
    console.log(`Preparando ${order.drink}...`)
    this.counter.handOver(order)
  }
  // endregion
}

class Counter {
  // region: Counter#hand_over
  handOver(order: Order) {
    console.log(`¡${order.drink} listo!`)
  }
  // endregion
}

// region: wiring
// Cada `new` es un cable: así se ve el circuito en código.
const taker = new OrderTaker(new Cashier(new Barista(new Counter())))
taker.take('espresso')
// endregion

File: src/levels/L00-tutorial/repaired.rb
Order = Struct.new(:drink)

class OrderTaker
  def initialize(cashier) = @cashier = cashier

  # region: OrderTaker#take
  def take(drink)
    @cashier.charge(Order.new(drink))
  end
  # endregion
end

class Cashier
  def initialize(barista) = @barista = barista

  # region: Cashier#charge
  def charge(order)
    puts "Cobrado: #{order.drink}"
    @barista.prepare(order)
  end
  # endregion
end

class Barista
  def initialize(counter) = @counter = counter

  # region: Barista#prepare
  def prepare(order)
    puts "Preparando #{order.drink}..."
    @counter.hand_over(order)
  end
  # endregion
end

class Counter
  # region: Counter#hand_over
  def hand_over(order)
    puts "¡#{order.drink} listo!"
  end
  # endregion
end

# region: wiring
# Cada `new` es un cable: así se ve el circuito en código.
taker = OrderTaker.new(Cashier.new(Barista.new(Counter.new)))
taker.take(:espresso)
# endregion
